patches/coreboot-4.8.1/0009-Add-heads-TPM-measurements-to-Skylake-Kabylake.patch

```diff
@@ -0,0 +1,181 @@
++			!IS_ENABLED(CONFIG_MEASURED_BOOT))
+ 		init_tpm(s3wake);
+ }
+ 
+@@ -483,8 +484,33 @@ void fsp_memory_init(bool s3wake)
+ 	if (status != CB_SUCCESS)
+ 		die("Loading FSPM failed!\n");
+ 
++	if (IS_ENABLED(CONFIG_MEASURED_BOOT) && IS_ENABLED(CONFIG_LPC_TPM)) {
++		// we don't know if we are coming out of a resume
++		// at this point, but want to setup the tpm ASAP
++		init_tpm(0);
++		tlcl_lib_init();
++		const void * const bootblock = (const void*) 0xFFFFF800;
++		const unsigned bootblock_size = 0x800;
++		tlcl_measure(0, bootblock, bootblock_size);
++
++		tlcl_measure(1, _romstage, _eromstage - _romstage);
++	}
++
++	if (IS_ENABLED(CONFIG_MEASURED_BOOT)) {
++	  tlcl_measure(1, (const void*) hdr.image_base, hdr.image_size);
++	}
++
+ 	/* Signal that FSP component has been loaded. */
+-	prog_segment_loaded(hdr.image_base, hdr.image_size, SEG_FINAL);
++	// Don't measure since it is relocated at this point
++	prog_segment_loaded(hdr.image_base, hdr.image_size, SEG_FINAL | SEG_NO_MEASURE);
+ 
+ 	do_fsp_memory_init(&hdr, s3wake, &memmap);
+ }
++
++void platform_segment_loaded(uintptr_t start, size_t size, int flags)
++{
++	if (IS_ENABLED(CONFIG_MEASURED_BOOT) && !(flags & SEG_NO_MEASURE)) {
++	  tlcl_measure(1, (const void*) start, size);
++	}
++}
++
+diff --git ./src/drivers/intel/fsp2_0/silicon_init.c ./src/drivers/intel/fsp2_0/silicon_init.c
+index bda88d1..49220af 100644
+--- ./src/drivers/intel/fsp2_0/silicon_init.c
++++ ./src/drivers/intel/fsp2_0/silicon_init.c
+@@ -18,6 +18,7 @@
+ #include <fsp/api.h>
+ #include <fsp/util.h>
+ #include <program_loading.h>
++#include <security/tpm/tss.h>
+ #include <stage_cache.h>
+ #include <string.h>
+ #include <timestamp.h>
+@@ -101,6 +102,10 @@ void fsps_load(bool s3wake)
+ 	if (rdev_readat(&rdev, dest, 0, size) < 0)
+ 		die("Failed to read FSPS!\n");
+ 
++	if (IS_ENABLED(CONFIG_MEASURED_BOOT)) {
++	  tlcl_measure(1, (const void*) dest, size);
++	}
++
+ 	if (fsp_component_relocate((uintptr_t)dest, dest, size) < 0)
+ 		die("Unable to relocate FSPS!\n");
+ 
+@@ -115,7 +120,7 @@ void fsps_load(bool s3wake)
+ 	stage_cache_add(STAGE_REFCODE, &fsps);
+ 
+ 	/* Signal that FSP component has been loaded. */
+-	prog_segment_loaded(hdr->image_base, hdr->image_size, SEG_FINAL);
++	prog_segment_loaded(hdr->image_base, hdr->image_size, SEG_FINAL | SEG_NO_MEASURE);
+ 	load_done = 1;
+ }
+ 
+diff --git ./src/drivers/intel/gma/opregion.c ./src/drivers/intel/gma/opregion.c
+index 70cbccc..e2cad8f 100644
+--- ./src/drivers/intel/gma/opregion.c
++++ ./src/drivers/intel/gma/opregion.c
+@@ -24,6 +24,7 @@
+ #include <device/pci_ops.h>
+ #include <console/console.h>
+ #include <cbmem.h>
++#include <security/tpm/tss.h>
+ #include "intel_bios.h"
+ #include "opregion.h"
+ 
+@@ -57,6 +58,10 @@ void *locate_vbt(size_t *vbt_size)
+ 	if (vbt_size)
+ 		*vbt_size = file_size;
+ 
++	if (IS_ENABLED(CONFIG_MEASURED_BOOT)) {
++	  tlcl_measure(2, (const void*) vbt_data, file_size);
++	}
```

Comment:
There is no need to do this here either because the vbt is loaded using cbfs_boot_load_file which itself calls cbfs_load_and_decompress which itself calls (via the measurements patch) prog_segment_loaded, which means you're causing this to be measured twice.

Reply:
Yup - my mistake. Removing